vocabularies.md

```diff
@@ -0,0 +1,283 @@
+          }
+        }
+      },
+      {
+        "id": "common_types",
+        "name": "common_types",
+        "scrollable": true,
+        "hierarchical": false,
+        "type": "vocabulary",
+        "_links": {
+          "entries": {
+            "href": "https://dspace7.4science.cloud/server/api/submission/vocabularies/srsc/entries"
+          },
+          "self": {
+            "href": "https://dspace7.4science.cloud/server/api/submission/vocabularies/srsc"
+          }
+        }
+      },
+      {
+        "id": "common_iso_languages",
+        "name": "common_iso_languages",
+        "scrollable": true,
+        "hierarchical": false,
+        "type": "vocabulary",
+        "_links": {
+          "entries": {
+            "href": "https://dspace7.4science.cloud/server/api/submission/vocabularies/common_iso_languages/entries"
+          },
+          "self": {
+            "href": "https://dspace7.4science.cloud/server/api/submission/vocabularies/common_iso_languages"
+          }
+        }
+      }
+    ]
+  },
+  "_links": {
+    "self": {
+      "href": "https://dspace7.4science.cloud/server/api/submission/vocabularies"
+    }
+  },
+  "page": {
+    "size": 20,
+    "totalElements": 3,
+    "totalPages": 1,
+    "number": 0
+  }
+}
+```
+
+## Single Vocabulary
+**/api/submission/vocabularies/<:vocabulary-name>**
+
+Provide detailed information about a specific controlled vocabulary. The JSON response document is as follow
+```json
+{
+  "id": "srsc",
+  "name": "srsc",
+  "scrollable": false,
+  "hierarchical": true,
+  "preloadLevel": 1,
+  "type": "vocabulary"
+}
+```
+
+Attributes
+* id: the id of the vocabulary, it is the same than the name
+* name: see id
+* scrollable: if true mean that it is possible to scroll all the entries in the authority without providing a filter parameter, see (vocabulary entries)[vocabularies.md#vocabulary-entries]
+* hierarchical: if true means that the vocabulary expose a tree structure where some entries are parent of others
+* preloadLevel: for hierarchical vocabularies express the preference to preload the tree at a specific level of depth (0 only the top nodes are shown, 1 also their children are preloaded and so on)
+
+Exposed links:
+* entries: get entries from the controlled vocabulary
+
+## Linked entities
+### controlled vocabulary entries
+**/api/submission/vocabularies/<:vocabulary-name>/entries**
+
+It returns the entries in the vocabulary in response to either a user input or a scrollable list, see below 
+
+The supported parameters are:
+* page, size [see pagination](README.md#Pagination)
+* metadata: the metadata field name (e.g. "dc.type") for which the controlled vocabulary is used: mandatory. The entries returned by a vocabulary may be different for different metadata fields.
+* collection: the uuid of the collection where the item is being submitted: mandatory. The entries returned by a vocabulary may differ per collection based on your submission configuration.
```

Comment:
When can they be different for separate collections fields but still using the same vocabulary ID? Can you provide a sample?
Otherwise I'd recommend to remove this parameter

Reply:
dc.type can be configured as a controlled vocabulary in the submission form for a specific collection (form definition) and as free text for another submission. Moreover, the collection parameter is part of the choice interface so some implementation could also return different entries depending on it similar to how the SOLRAuthority is currently using the metadata parameter

Reply:
Based on https://github.com/DSpace/Rest7Contract/pull/128#discussion_r443391928, I assume this currently also only returns different entries for the same vocabulary-name for solr. If that example is included as I requested in https://github.com/DSpace/Rest7Contract/pull/128#discussion_r443629083, it should be much clearer

Reply:
@benbosman : I had the same confusion over the `collection` endpoint initially until I looked at the current (DSpace 6 and below) code implementation.  It *is possible* to configure in input-forms.xml (now submission-forms.xml) a value-pairs that is *collection specific* (as any submission field can be made to only appear for specific collections, or even have different configurations for different collections -- e.g. one collection could have a "dc.type" that uses value-pairs, while in another "dc.type" can be configured to be free text, or a _different set_ of value-pairs).

You'll see in this [prior discussion](https://github.com/DSpace/Rest7Contract/pull/128#discussion_r439499864) I was the one who suggested the above wording.  So, I think that, for now *both metadata and collection* must be required to work with our current code (especially for value-pairs). In the future though, I fully agree this is a bit odd, and we may want to revisit this in a future release (not 7.0, more like 8.0 unless we find a need to change it earlier).

Reply:
I'd recommend to add in at least some detail on this line explaining this is currently not used for any implementation, but it's possible to create different implementations per collection
The difference in value pairs per collection are determined by the vocabulary-name, not by the collection UUID
Value pairs don't differ per metadata field, only per vocabulary-name
That should be a really small change, and this is very obviously an unclear area

Reply:
@benbosman : Good point that `value-pairs` actually don't require a collection UUID to look them up. They have a unique vocabulary name.  With regards to this `collection` param, I've just added an inline question on the implementation PR for @abollini : https://github.com/DSpace/DSpace/pull/2792#pullrequestreview-438617390

The more I look at the implementation PR, the more I'm suspecting this `collection` param doesn't need to be required (and we may not even need it at all).  That said, as noted above, the `metadata` param definitely needs to be required for `SolrAuthority` (at a minimum).

Reply:
the `collection` and `metadata` params come mostly from the authority control framework existing methods signature that include such parameters see for instance https://github.com/DSpace/DSpace/blob/30ad7676ef92df507cffb7a2040f5ef3ee5e56b9/dspace-api/src/main/java/org/dspace/content/authority/ChoiceAuthorityServiceImpl.java#L106

The other scenario where these parameters are needed is for consistency check, if we only have the metadata we cannot know for sure that it is associated with a specific vocabulary. The dc.type can be associated for some collection to common_types, for other to "other_list_types" and for other ones can be a free text. In the edge scenario this lead to inconsistent request as reported here https://jira.lyrasis.org/browse/DS-4371

As said in https://github.com/DSpace/Rest7Contract/pull/128#discussion_r437661208 I'm in favor to remove **both** parameters but this will require additional changes to the interfaces and a partial refactor of the SolrAuthority that is actually the only ChoiceAuthority where, at least the metadata parameter is really used.
It should be not too complex, as it could be enough to add the named plugin interface to the authority so to retrieve additional configuration for the plugin specific for the assigned name. This will allow to have multiple instance of the SolrAuthority class one for the one metadata and one for another, etc.
I'm willing to explore this further refactoring if there is an agreement on that I just want to avoid the risk to make a single PR bigger and bigger.

Reply:
Thanks for pointing this out, https://jira.lyrasis.org/browse/DS-4371 is indeed another example clarifying how the mandatory metadata and collection fields can be misinterpreted, this time the vocabulary was determined based on the parameters instead of the vocabulary name.

Perhaps the parameters should only be set mandatory for authorities (a smaller change compared to completely removing them), since the others are not supposed to use them. That makes their use clearer. Angular is of course still free to include them for non-authorities, assuming REST will ignore them.
By using that in-between step, it's also easier to redesign the authority core to expect a separate name in e.g. 7.1 and remove the parameters completely

Reply:
@abollini : I'm OK with @benbosman's suggestion here.  If it's easier for now to just make these params _only mandatory for authorities_, then that seems like a step in the right direction.  I also agree we can just have the Angular team pass these all the time (as long as they are easy to achieve), but in the REST API they can be ignored except when needed for authorities.

Reply:
unfortunately, I don't see that as a simplification nor as a step forward the right solution. All the three things value-pairs, vocabularies and authorities need to be managed in the same way. Add an exception here is not the way to go imho. If we don't want to remove the collection/metadata from the authority interfaces we need to keep them also for the other "things" making sure that they are used in a consistent ways (i.e. they match with the specified vocabulary name to avoid the issue DS-4371. This is exactly what is done here in the current implementation https://github.com/DSpace/DSpace/pull/2792/files#diff-151c8c35ac6f56dc3482b82c56b0474cR85 see also the corresponding test https://github.com/DSpace/DSpace/pull/2792/files#diff-379fecc95ee8bd1f6f91349f0cd7e9d3R193-R206  ). Anyway, **I'm trying to implement the change to remove completely these params** in a parallel branch so that you can check exactly how much changes are needed (it seems less than what I initially suspected)

Reply:
here the result of my investigation: https://github.com/4Science/DSpace/pull/157 it looks to me as a better, cleaner approac. Do you agree? should I remove completely the metadata & collection parameters? if you like that I will apply the related change also to this contract https://github.com/4Science/DSpace/pull/157

Reply:
Yes, since that change looks simple enough. I'd rather we just remove the metadata/collection parameters.  If the three of us are confused by them, they will be even more confusing to others. So, if we can easily remove them now, I'm in favor of doing so.